src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/Level.java

```diff
@@ -46,47 +95,117 @@ private Level()
+        nodeCenterList.add(ladderNode.getNorthNodeCenter());
+        nodeCenterList.add(ladderNode.getSouthNodeCenter());
+        nodeCenterList.add(ladderNode.getEastNodeCenter());
+        nodeCenterList.add(ladderNode.getWesthNodeCenter());
+
+        for(final Point2D.Double pos: nodeCenterList)
         {
-            ladderNode.setDirectionNegX(Node.NodeStatus.LADDER);
-            cobbleNode.setDirectionPosX(Node.NodeStatus.LADDER);
+            if(cobbleCenter.equals(pos) || ladderCenter.equals(pos))
+            {
+                continue;
+            }
+            final Node tempNode = new Node(pos.getX(), pos.getY(), ladderCenter);
+            tempNode.setStyle(TUNNEL);
+            nodes.put(pos, tempNode);
+            openNodes.add(tempNode);
         }
-        else if (buildingMiner.getVectorX() < 0)
+    }
+
+    /**
+     * Getter for a random Node in the level.
+     * @return any random node.
+     */
+    public Node getRandomNode()
+    {
+        return openNodes.peek();
+    }
+
+    /**
+     * Closes the first Node in the list (Has been returned previously probably).
+     * Then creates the new nodes connected to it.
+     * @param rotation
+     */
+    public void closeNextNode(int rotation)
+    {
+        final Node tempNode = openNodes.poll();
+        final List<Point2D.Double> nodeCenterList = new ArrayList<>();
+
+        switch(tempNode.getStyle())
         {
-            ladderNode.setDirectionPosX(Node.NodeStatus.LADDER);
-            cobbleNode.setDirectionNegX(Node.NodeStatus.LADDER);
+            case TUNNEL:
+                nodeCenterList.add(getNextNodePositionFromNodeWithRotation(tempNode, rotation, 0));
+                break;
+            case BEND:
+                nodeCenterList.add(getNextNodePositionFromNodeWithRotation(tempNode, rotation, ROTATE_THREE_TIMES));
+                break;
+            case CROSSROAD:
+                nodeCenterList.add(getNextNodePositionFromNodeWithRotation(tempNode, rotation, 0));
+                nodeCenterList.add(getNextNodePositionFromNodeWithRotation(tempNode, rotation, ROTATE_ONCE));
+                nodeCenterList.add(getNextNodePositionFromNodeWithRotation(tempNode, rotation, ROTATE_THREE_TIMES));
+                break;
+            default:
+                return;
         }
-        else if (buildingMiner.getVectorZ() > 0)
+
+        for(final Point2D.Double pos: nodeCenterList)
         {
-            ladderNode.setDirectionNegZ(Node.NodeStatus.LADDER);
-            cobbleNode.setDirectionPosZ(Node.NodeStatus.LADDER);
+            if(nodes.containsKey(pos))
+            {
+                continue;
+            }
+            final Node tempNodeToAdd = new Node(pos.getX(), pos.getY(), new Point2D.Double(tempNode.getX(), tempNode.getZ()));
+            int randNumber = rand.nextInt(RANDOM_TYPES);
+            tempNodeToAdd.setStyle(randNumber <= 1 ? Node.NodeType.TUNNEL : (randNumber == 2 ? Node.NodeType.BEND : Node.NodeType.CROSSROAD));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'Node.NodeType.TUNNEL' due to existing static import 'com.minecolonies.coremod.entity.ai.citizen.miner.Node.NodeType.TUNNEL' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)